injection/informers.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package injection
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"k8s.io/client-go/rest"
+
+	"github.com/knative/pkg/controller"
+)
+
+// InformerInjector holds the type of a callback that attaches a particular
+// informer type to a context.
+type InformerInjector func(context.Context) (context.Context, controller.Informer)
+
+var (
+	informerMutex sync.RWMutex
+	informers     []InformerInjector
+)
+
+func RegisterInformer(ii InformerInjector) {
+	informerMutex.Lock()
+	defer informerMutex.Unlock()
+
+	informers = append(informers, ii)
+}
+
+func GetInformers() []InformerInjector {
+	informerMutex.RLock()
+	defer informerMutex.RUnlock()
+
+	return informers[:]
+}
+
+func SetupInformers(ctx context.Context, cfg *rest.Config, resyncPeriod time.Duration) (context.Context, []controller.Informer) {
+	// Based on the reconcilers we have linked, build up a set of clients and inject
+	// them onto the context.
+	for _, ci := range GetClients() {
+		ctx = ci(ctx, cfg)
+	}
+
+	// Based on the reconcilers we have linked, build up a set of informer factories
+	// and inject them onto the context.
+	for _, ifi := range GetInformerFactories() {
+		ctx = ifi(ctx, resyncPeriod)
```

Comment:
I think we want `resyncPeriod` on `context.Context`, so that the tracker can key off of the same value.